### DISCUSSION TITLE
customFunction module as reaction coordinate

### DISCUSSION THREAD

**Discussion by satyajitkhatua09** [Colvars/colvars#690] [Q&A] [answered]
Hi,

This is Satyajit Khatua. I am here to gather some expert opinions about studying the conformational dynamics of DNA unwinding. To capture both the rotation and the translocation of helices, I have to use the combination of some colvars that capture the change in distance and rotation around a helical axis. 

Do any of you have any idea about the linear combination of any two such colvars (like distance/diatanceZ and spinAngle) to implement inside the customFunction colvar that can capture the conformational change in DNA in terms of simultaneous rotation and translocation?

Best,
Satyajit Khatua

**Comment by giacomofiorin** [comment, marked as answer]
Hi Satyajit!

To use a linear combination you could use `componentCoeff`: 
https://colvars.github.io/master/colvars-refman-namd.html#sec:cvc_superp
The example below defines a colvar that is the average of two functions:
```
colvar {
    name xi
    function1 {
        componentCoeff 0.5
        ....
    }
    function2 {
        componentCoeff 0.5
        ....
    }
}
```

`customFunction` is much more flexible, because you can use many more mathematical functions, and I do recommend using it. The only caveat is that GROMACS and VMD currently would not support it.

If you have a specific idea in mind, please describe it so that we can confirm whether it's supported or not.

Cheers!

**Reply by satyajitkhatua09** [reply]
Hi @giacomofiorin,

I will give it a try. One quick question, inside function brackets I have to put the colvar module as is defined in the manual, right?

Best,
Satyajit

**Reply by giacomofiorin** [reply]
Based on what you mentioned, function1 would be `distanceZ` and function2 would be `spinAngle`. Each block that follows them would have their usual parameters, plus the coefficient for linear superposition.

If you stick with linear superposition, I would definitely recommend also using the Dashboard in VMD. 

If you use `customFunction`, the Dashboard would be less useful because, sadly, the VMD developers would not allow us to include the Lepton library, which is needed for that feature.

**Reply by satyajitkhatua09** [reply]
Hi @giacomofiorin,

Here is the snippet of my configuration file where I have turned the collar module on.

**# Incorporating the colvar file
colvars                      on;                 # turing collective variable (CV) module on
colvarsConfig           colvars.in;         # file containing the CV information**

Inside the colvars.in file, I have defined the two CVs as follows:

colvarsTrajFrequency     100
colvarsRestartFrequency  500

colvar {
  name xi

  # using the coefficient for linear superposition

  function1 {
    componentCoeff 0.5
    distanceZ {
      main {
        atomsFile        atoms.pdb
        atomsCol         B
        atomsColValue    1.0
      }
      ref {
        atomNameResidueRange CA 3778-3956
      }
      ref2 {
        atomNameResidueRange CA 408-623
      }
    }
  }
  function2 {
    componentCoeff 0.5
    spinAngle {
      atoms {
        atomsFile            atoms.pdb
        atomsCol            B
        atomsColValue   1.0
      }
      refPositionsFile     ref_pos_inpcrd.pdb
      axis                       (-0.5, 0.3, 0.8)
    }
  }
}

However, when I ran the simulation, it threw the error message attached below. I am unable to understand the real problem here that reflects such behavior.

colvars: ----------------------------------------------------------------------
colvars: Initializing the collective variables module, version "2020-07-07".
colvars: Please cite Fiorin et al, Mol Phys 2013:
colvars:  https://dx.doi.org/10.1080/00268976.2013.813594
colvars: in any publication based on this calculation.
colvars: SMP parallelism is enabled; if needed, use "smp off" to override this.
colvars: This version was built with the C++11 standard or higher.
colvars: Using NAMD interface, version "2020-05-04".
colvars: ----------------------------------------------------------------------
colvars: Reading new configuration from file "colvars.in":
colvars: # units = "" [default]
colvars: # smp = on [default]
colvars: # colvarsTrajFrequency = 100
colvars: # colvarsRestartFrequency = 500
colvars: # scriptedColvarForces = off [default]
colvars: # scriptingAfterBiases = off [default]
colvars: ----------------------------------------------------------------------
colvars:   Initializing a new collective variable.
colvars:   # name = "xi"
colvars: Error: no valid components were provided for this collective variable.
FATAL ERROR: Error in the collective variables module (see above for details)
[Partition 0][Node 0] End of program

**Reply by giacomofiorin** [reply]
"function1" is only an example, which in your case would be distanceZ

**Reply by satyajitkhatua09** [reply]
Hi @giacomofiorin,

Thank you, now it is running perfectly.

So, if I want to use the _customFunction_ option, do I have to first manually deduce the functional form that co-relates the two CV modules?

**Comment by jhenin** [comment]
Hi @satyajitkhatua09 ,
to add to Giacomo's advice: if you already have example structures for the end points (or simulation trajectories), you can load them in VMD and experiment freely with variables using the [Colvars Dashboar](https://colvars.github.io/master/colvars-refman-vmd.html#sec:dashboard)d, until you find one that suits your needs.
Jérôme

**Reply by satyajitkhatua09** [reply]
Hi @jhenin,

Thank you for letting me know the feature. It is very useful to define the colvar modules.

Best,
Satyajit
